app/src/main/java/org/simple/clinic/analytics/TimingCollector.kt

```diff
@@ -0,0 +1,98 @@
+package org.simple.clinic.analytics
+
+import androidx.annotation.VisibleForTesting
+import org.simple.clinic.util.UtcClock
+import org.threeten.bp.Duration
+import org.threeten.bp.Instant
+import java.util.UUID
+import java.util.concurrent.atomic.AtomicBoolean
+
+/**
+ * This class is meant to collect timing information for an operation over time. An operation can
+ * have multiple stages, and the collector allows collecting discrete timing information for each
+ * of these stages.
+ *
+ * Every instance of the collector is given a run ID, so that all events from a given run can be
+ * identified and grouped together if required.
+ *
+ * ### Example Usage
+ * ```
+ * val collector = TimingCollector("Search for patients", utcClock)
+ *
+ * collector.start("fetch-patient-names")
+ * // Fetch patient names from database
+ * collector.stop("fetch-patient-names")
+ *
+ * collector.start("fuzzy-name-matching")
+ * // Use fuzzy name matching to drop patients that don't match  criteria
+ * collector.stop("fuzzy-name-matching")
+ *
+ * collector.start("fetch-patient-details")
+ * // Fetch complete patient details from database
+ * collector.stop("fetch-patient-details")
+ *
+ * collector.start("sort-by-facility")
+ * // Sort the final search results list according the patients visited facility list
+ * collector.stop("stop-by-facility")
+ *
+ * // Once the entire operation is complete, push higher level events to analytics
+ * collector.finish()
+ * ```
+ *
+ * **Note:** Since an operation can happen across multiple threads, this class has been designed
+ * with thread-safety in mind. So it uses immutable maps for tracking operations, and atomic
+ * primitives where necessary.
+ **/
+class TimingCollector(
+    val name: String,
+    private val clock: UtcClock
+) {
+  @get:VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+  var ongoing: Map<String, Instant> = emptyMap()
+    private set
+
+  @get:VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+  var completed: Map<String, Duration> = emptyMap()
+    private set
+
+  val runId: String = UUID.randomUUID().toString()
+
+  private val finished = AtomicBoolean(false)
+
+  fun start(stageName: String) {
+    if(finished.get().not()) {
+      ongoing += "$name:$stageName" to Instant.now(clock)
+    }
+  }
+
+  fun stop(stageName: String) {
+    if(finished.get().not()) {
+      val analyticsEventName = "$name:$stageName"
+      ongoing[analyticsEventName]?.let { stageStartedAt ->
+        val now = Instant.now(clock)
+        val timeTakenToCompleteStage = Duration.between(stageStartedAt, now)
```

Comment:
`Duration.between()` can return negative values.
Should we calculate absolute value here?

Reply:
Good catch. I'll change it to an absolute duration.